compiler/src/iree/compiler/Codegen/Common/GPU/DecomposeHorizontallyFusedGemms.cpp

```diff
@@ -0,0 +1,245 @@
+
+          IRMapping regionMapping;
+
+          for (auto [oldBlockArgNum, newBlockArg] :
+               llvm::zip_equal(usedInputs, blockArgs)) {
+            regionMapping.map(oldBody->getArgument(oldBlockArgNum),
+                              newBlockArg);
+          }
+
+          for (Operation *usedOperation : usedOperations) {
+            b.clone(*usedOperation, regionMapping);
+          }
+
+          b.create<linalg::YieldOp>(loc, regionMapping.lookup(result));
+        });
+
+    if (unusedDims.none()) {
+      DictionaryAttr origDictAttr = linalgOp->getDiscardableAttrDictionary();
+      auto loweringConfigAttr = dyn_cast_or_null<IREE::GPU::LoweringConfigAttr>(
+          origDictAttr.get(IREE::GPU::LoweringConfigAttr::getMnemonic()));
+      DictionaryAttr newDictAttr;
+      if (loweringConfigAttr &&
+          loweringConfigAttr.getAttributes().get(
+              IREE::GPU::getPromotedOperandListAttrName())) {
+        SmallVector<NamedAttribute> newAttrList;
+        for (auto origAttr : origDictAttr.getValue()) {
+          if (origAttr.getName() !=
+              IREE::GPU::LoweringConfigAttr::getMnemonic()) {
+            newAttrList.push_back(origAttr);
+            continue;
+          }
+          SmallVector<unsigned> operandNums =
+              llvm::map_to_vector(inputs, [](OpOperand *operand) {
+                return operand->getOperandNumber();
+              });
+          auto range = llvm::map_range(inits, [](OpOperand *operand) {
+            return operand->getOperandNumber();
+          });
+          operandNums.append(range.begin(), range.end());
+          auto origGPUAttr =
+              cast<IREE::GPU::LoweringConfigAttr>(origAttr.getValue());
+          IREE::GPU::LoweringConfigAttr newGPUAttr =
+              getModifiedLoweringConfigForDecomposedGemmOp(
+                  rewriter, origGPUAttr, operandNums);
+          newAttrList.emplace_back(
+              NamedAttribute{origAttr.getName(), newGPUAttr});
+        }
+        newDictAttr = DictionaryAttr::get(rewriter.getContext(), newAttrList);
+      } else {
+        newDictAttr = origDictAttr;
+      }
+      newOp->setDiscardableAttrs(newDictAttr);
+    }
+
+    rewriter.replaceAllUsesWith(linalgOp->getResult(resultNumber),
+                                newOp.getResult(0));
+  }
+
+  rewriter.eraseOp(linalgOp);
+  return success();
+}
+
+void DecomposeHorizontallyFusedGemmsPass::runOnOperation() {
+  auto funcOp = getOperation();
+  IRRewriter rewriter(&getContext());
+  SmallVector<linalg::LinalgOp> horizontallyFusedOps;
+  funcOp.walk([&](linalg::LinalgOp linalgOp) {
+    if (IREE::LinalgExt::isaHorizontallyFusedContraction(linalgOp)) {
+      horizontallyFusedOps.push_back(linalgOp);
+    }
+  });
+
+  for (auto linalgOp : llvm::make_early_inc_range(horizontallyFusedOps)) {
+    if (failed(decomposeHorizontallyFusedGemmOperations(rewriter, linalgOp))) {
+      return signalPassFailure();
+    }
```

Comment:
FYI, the Walk Pattern Rewrite Driver would work for this use case: https://mlir.llvm.org/docs/PatternRewriter/#walk-pattern-rewrite-driver

Reply:
I dont want to use pattern rewriter here. So ill just do a walk.